<a href="https://colab.research.google.com/github/iestleija/5094-Matematicas/blob/Unidad3/Unidad3/Practices/5094_U3_C3_266229.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Maestría en Inteligencia Artificial y Analítica de Datos

- Curso: **Matemáticas y estadística para inteligencia artificial**

- Semestre: Enero-Junio

- Profesor: Dra. Helen Clara Peñate Rodríguez
- Actividad: **C3.Probabilidad**
- Alumno: Ismael Estrada Leija
- Matrícula: 266229

# Actividad. C3.Probabilidad
Enlace a colab: https://colab.research.google.com/drive/1stIGQMnEfNWDypq3Ciz3mNqqKVANInJY?usp=sharing

# Instrucciones
## Probabilidades

# 1. Modelo de detección de fraude y decisión operativa

Una empresa de pagos digitales utiliza un modelo de inteligencia artificial para detectar transacciones fraudulentas.

Se sabe que:
- El 1.5% de las transacciones son realmente fraudulentas.
- El modelo detecta correctamente una transacción fraudulenta el 96% de las veces.
- El modelo marca como fraudulenta una transacción legítima el 4% de las veces.


Definamos

- $F$ como el caso de una operación fraudulenta comprobada
- $M$ como el caso de una marcación positiva de fraude como salida del modelo.

Entonces,

$P(F)=1.5\%=0.015$
por lo tanto,

$P(\neg F)=98.5\%=0.985$

$P(M | F) = 96\%$ y,

$P(M | \neg F) = 4\%$


In [70]:
p_fraude = 0.015
p_no_fraude = 1-p_fraude
p_marcar_dado_fraude = 0.96
p_marcar_no_dado_fraude = 1-p_marcar_dado_fraude

### Ejercicios 1

## (a) Calcule la probabilidad de que una transacción sea marcada como fraude por el modelo.
---

Deseamos conocer la probabilidad total de marcación de fraude $P(M)$, para lo cual la fórmula de la probabilidad total de acuerdo a taro **** es:

$$P(A) = Σ (A|H_i)*P(H_i)$$

Dónde $H_i$ son sucesos que son particiones del espacio muestral y lo cubren en su totalidad.

Para nuestro caso la fórmula queda como:

$$P(M) = P(M|F)* P(F) + P(M| \neg F)*P(\neg F)$$

Considerando que el espacio muestral esta cubierto por las particiones $F$ y $\neg F$

In [72]:
# calculemos...
p_total_marcadas = p_marcar_dado_fraude*p_fraude + p_marcar_no_dado_fraude*p_no_fraude
print(f"La probabilidad total de marcación por el modelo es: {p_total_marcadas:.4f}");

La probabilidad total de marcación por el modelo es: 0.0538


## (b) Si una transacción fue marcada como fraude, calcule la probabilidad de que realmente sea fraudulenta.
---
Utilizando el teorema o regla de Bayes determinaremos la probabilidad de que una transaccion marcada sea realmente fraudulenta; recordemos que nuestro espacio muestral fue particionado por las probabilidades complementarias de que una transaccion sea fraudulenta o no. Se sobrepone en ese contexto la probabilidad de que una transacción sea marcada por el modelo con intersecciones en las particiones iniciales.
$$P(F|M) = \frac {P(M|F)*P(F)}{P(M)}$$
Nuestro espacio muestral en este caso se limita a las transacciones marcadas y queremos saber la razón de aquellas que son realmente fraudulentas sobre el total marcadas.


In [73]:
# Probabilidad de Fraude positivo (real)
p_fraude_real =  (p_marcar_dado_fraude * p_fraude) / p_total_marcadas
print(f"La probabilidad de marcar fraudes reales es: {p_fraude_real:.4f}")

La probabilidad de marcar fraudes reales es: 0.2677


## (c) Si una transacción no fue marcada como fraude, calcule la probabilidad de que realmente sea fraudulenta.
La partición de las transacciones no marcadas como fraude es la probabilidad complementaria de las marcadas $P(\neg M) = 1 - P(M)$ y aplicaremos la fórmula previamente utilizada
$$P(F|\neg M) = \frac {P(\neg M|F)*P(F)}{P(\neg M)}$$


In [79]:
p_total_no_marcadas = 1 - p_total_marcadas
p_no_marcar_dado_fraude = 1-p_marcar_dado_fraude

p_fraude_no_marcado = (p_no_marcar_dado_fraude * p_fraude) / p_total_no_marcadas;
print(f"La probabilidad de que una transacción sea fraudulenta y no haya sido marcada es de :{p_fraude_no_marcado} ")

La probabilidad de que una transacción sea fraudulenta y no haya sido marcada es de :0.0006341154090044393 


## (d) Calcule la probabilidad de que una transacción sea legítima dado que fue marcada como fraude.
---
Nuevamente nuestro espacio muestral se limita a las transacciones marcadas y ahora calcularemos la probabilidad complementaria para el caso de falsos positivos.
$$P(\neg F|M) = \frac {P(M|\neg F)*P(\neg F)}{P(M)}$$


In [75]:
p_falso_positivo =  (p_marcar_no_dado_fraude * p_no_fraude) / p_total_marcadas
print(f"La probabilidad de que una transacción sea legítima dado que fue marcada como fraude es: {p_falso_positivo:.4f}");

La probabilidad de que una transacción sea legítima dado que fue marcada como fraude es: 0.7323


## (e) Si la empresa procesa 200,000 transacciones diarias:
  - i. ¿Cuántas transacciones se espera que sean marcadas como fraude?


In [76]:
cant_transacciones = 200000
total_marcadas = p_total_marcadas * cant_transacciones
print(f"Se espera que sean marcadas como fraude: {total_marcadas:.0f}");

Se espera que sean marcadas como fraude: 10760


  - ii. ¿Cuántas de las marcadas como fraude serán realmente fraudulentas?

In [77]:
fraudulentas_marcadas = p_fraude_real * total_marcadas
print(f"De las marcadas como fraude serán realmente fraudulentas: {fraudulentas_marcadas:.0f} equivalente al {p_fraude_real:.2%} de las marcadas");

De las marcadas como fraude serán realmente fraudulentas: 2880 equivalente al 26.77% de las marcadas


  - iii. ¿Cuántas serán falsas alarmas?

In [78]:
legitimas_marcadas = p_falso_positivo * total_marcadas
print(f"De las marcadas como fraude serán falsas alarmas: {legitimas_marcadas:.0f} equivalente al {p_falso_positivo:.2%} de las marcadas");

De las marcadas como fraude serán falsas alarmas: 7880 equivalente al 73.23% de las marcadas


## (f) Desde el punto de vista empresarial, explique el riesgo de evaluar el modelo únicamente con el porcentaje de detección correcta de fraudes.
---
Dado que el porcentaje de casos fraudulentos es muy bajo, la alta precisión del modelo se ve reducida por su contraparte, o falsos positivos, que sera ponderada con una partición enorme del espacio muestral. Derivado de lo anterior en este caso tenemos que las marcas realizadas son de muy baja calidad presentando un 73.23% de falsas alarmas que resultan en fricción con los clientes, pérdida de confianza, trabajo de remediación, entre otros inconvenientes derivados del malestar de los consumidores.

El nivel de falsos positivos es altísimo para este caso en particular y no es aceptable.

| --Partición inicial-- | Cantidad | precisión/falla | Marcadas | Calidad |
| :--- | :---: | ---: | ---: | :---: |
| $P(F) = 1.5 \%$ | 3,000 |96% |2,880 | Fraude |
| $P(\neg F)=98.5 \%$ | 197,000 | 4% | 7,880 | Falsa Alarma|


**Sugerencia**: Primero calcule la probabilidad total de que el modelo marque una transacción como fraude combinando los casos en que la transacción es realmente fraudulenta y los casos en que no lo es. Luego utilice la definición de probabilidad condicional para responder las preguntas posteriores.

# 2. **Selección de variables y riesgo de sobreajuste**

Una empresa desarrolla un modelo de scoring crediticio.

Tiene 15 variables candidatas:

- 6 son realmente predictivas.
- 9 son ruido.


**Sugerencia**: Utilice combinaciones:
$$\begin{pmatrix} n \\ r \end{pmatrix} = \frac {n!}{r!(n-r!)}$$
y probabilidad clásica:
$$P(A) = \frac{casos \phantom{a} favorables}{casos \phantom{a} posibles}$$

Para la última parte, use independencia entre intentos y la distribución binomial.

# Ejercicios 2
Se seleccionan 5 variables al azar para entrenar un modelo preliminar.


## (a) ¿Cuál es la probabilidad de que el modelo incluya exactamente 3 variables realmente predictivas? = 23.98%
---
Definimos

$X$ = la función o variable aleatoria que retorna la cantidad de variables predictivas en una muestra de 5 variables.

Siguiendo la recomendación de usar la probabilidad clásica

$$P(X=3) = \frac{casos \phantom{a} favorables}{casos \phantom{a} posibles}$$

Calculemos los $casos$ $posibles$: la cantidad total de combinaciones en muestras de 5 variables. Con el coeficiente binomial podemos calcular el total de combinaciones de tamaño $r$ en una poblacion de tamaño $n$
$$\begin{pmatrix} n \\\\ r \end{pmatrix} = \frac {n!}{r!(n-r)!}$$
$$C_{totales}=\begin{pmatrix} 15 \\\\ 5 \end{pmatrix} =
= \frac {15!}{5!(15-5)!} = 3003$$

Para formar la muestra de cinco variables y cumplir con $P(X=3)$ se requieren 3 predictivas y 2 de ruido, las combinaciones para obtenerlas es
$$C_{predictivas} = \begin{pmatrix} 6 \\\\ 3 \end{pmatrix}
= \frac {6!}{3!(6-3)!} = 20$$

$$C_{ruido} = \begin{pmatrix} 9 \\\\ 2 \end{pmatrix}
= \frac {9!}{2!(9-2)!} = 36$$
De aqui se sigue que

$$P(X=3) = \frac{casos \phantom{a} favorables}{casos \phantom{a} posibles}
= \frac {C_{predictivas} * C_{ruido}}{C_{totales}} $$

$$P(X=3) = = \frac {20*36}{3003}=0.23976$$

## (b) ¿Cuál es la probabilidad de que incluya al menos 4 variables predictivas? = 0.199%
---
La probabilidad de al menos 4 variables predictivas es la siguiente y se detalla su cálculo enseguida
$$P(X>=4) = P(X=4) + P(X=5) = 0.046953047$$

---
Siguiendo los cálculos mostrados en el punto a) calculamos la $P(X=4)$ y $P(X=5)$

$$C_{predictivas} = \begin{pmatrix} 6 \\\\ 4 \end{pmatrix}  
= \frac {6!}{4!(6-4)!} = 15$$

$$C_{ruido} = \begin{pmatrix} 9 \\\\ 1 \end{pmatrix}
= \frac {9!}{1!(9-1)!} = 9$$

$$P(X=4) = \frac {C_{predictivas} * C_{ruido}}{C_{totales}} = \frac {15*9}{3003} = 0.044955045$$

$$C_{predictivas} = \begin{pmatrix} 6 \\\\ 5 \end{pmatrix}  
= \frac {6!}{5!(6-5)!} = 6$$

$$C_{ruido} = \begin{pmatrix} 9 \\\\ 0 \end{pmatrix}
= \frac {9!}{0!(9-0)!} = 1$$

$$P(X=5) = \frac {C_{predictivas} * C_{ruido}}{C_{totales}} = \frac {6*1}{3003} = 0.001998002$$

## (c) ¿Cuál es la probabilidad de que el modelo haya sido construido con mayoría de variables ruido (3 o más ruido)? = 71.33%
---
Para mantener consistencia $X$ sigue representando la cantidad de variables predictivas y siguiendo el método mostrado antes se calculan las probabilidades parciales.

$P(X=0) = 0.041958042$ / 5 de ruido

$P(X=1) = 0.251748252$ / 4 de ruido

$P(X=2) = 0.41958042$ / 3 de ruido

Por lo anterior, la probabilidad de que el modelo tenga 3 o mas variables de ruido es:

$P(X \leq 2) = P(X=0) + P(X=1) + P(X=2) = 0.713286713$




## (d) Si se repite este proceso 10 veces (selecciones independientes), ¿cuál es la probabilidad de que al menos uno de los modelos tenga 4 o más variables realmente predictivas? = 38.18%
---
Definamos $Y$ como la cantidad de modelos propuestos con al menos 4 variables predictivas despues de realizar 10 selecciones independientes.

$$b(x;n,p)=\begin{pmatrix} n \\\\ x \end{pmatrix}p^x(1-p)^{n-x}$$

Tenemos como premisa la probabilidad de al menos contar con 4 variables predictivas en una muestra, que ya calculamos

$$p = P(X \geq 4) = 0.046953$$
$$q = 1-p = 0.953047$$

Para el caso de "al menos uno" se recomienda calcular la probabilidad complementaria. Así, la probabilidad de 10 fracasos consecutivas es:

$$P(Y=0)= b(0;10, 0.953047)=\begin{pmatrix} 10 \\\\ 0 \end{pmatrix}p^0q^{10}=1*1*q^{10} = 0.61822$$

Por complemento, tenemos nuestra respuesta:
 $$P(Y \geq 1)=1-P(Y=0)= 1 - 0.61822 = 0.38178$$


## (e) Interprete los resultados en términos del riesgo de lanzar un modelo construido bajo selección aleatoria de features.
---
Bajo el contexto planteado, dentro de la fuente de variables, se cuenta con una cantidad importante de variables inútiles para la predicción, o ruido, a razón de 9/15.

Contando con un origen de variables en este estado y dejando al azar la selección de variables recibiremos un modelo con mas ruido que capacidad predictiva, cuando tienen menos de 3 variables predictivas, mostrado enseguida
$$P(X \leq 2) = 71.33\%$$

Aún con 10 intentos de selección,  la probabilidad de lograr un modelo con por lo menos 4 variables predictivas, es apenas del $38.18\%$

El tema del sobreajuste es ampliamente abordado en el libro de Gareth, (Gareth et al., 2023) y básicamente se refiere a que la existencia de variables que no se relacionan con el propósito del modelo generarán una falsa expectativa del rendimiento del mismo, por lo que se prefiere evaluar cuando se agregan variables el efecto en la precisión del modelo.



#Referencias
Gareth, J., Daniela, W., Trevor, H., Robert, T., & Jonathan, T. (2023). An Introduction to Statistical Learning.